### PR TITLE
docs(API) Add error report introduction of loaders (#2231)

### LINKED ISSUE #2231: Document webpack change: Add loader name to error message. Resolves #2878
*A pull request by @mc-zone was merged and maintainers requested a documentation change.*

See pull request: https://github.com/webpack/webpack/pull/6542

---

**What kind of change does this PR introduce?**
Feature.

**Did you add tests for your changes?**
Yes.

**If relevant, link to documentation update:**
N/A

**Summary**
Add loader name to Error message so that people can actually realize which loaders were emiting these errors clearly.

### Demo:

webpack.config.js:
```javascript
module.exports = {
  entry:"./index.js",
  output:{
    path:path.resolve(__dirname, "./dist"),
    filename:"[name].bundle.js",
  },
  mode:"development",
  module: {
    rules: [
      {
        test: /\.test$/,
        use: [
          "babel-loader",
          {
            loader:path.resolve(__dirname, './test-loader.js'),
          }
        ]
      },
    ]
  },
};
```

test-loader.js:
```javascript
module.exports = function(source){
  this.emitWarning(new Error("This is a emit Warning"));
  this.emitError(new Error("This is a emit Error"));
  return source;
};
```

lib.test:
```javascript
);a
console.log("I'm test");

```

### Current output:
image

### After this PR:
image

Resolves #2878.

**Does this PR introduce a breaking change?**
Nope (I think).

**Other information**
<del>Note: This feature also need some changes in `webpack/loader-runner` to handle errors from `throw` and `callback(error)` (need add loader name to `error.from`). I will add it subsequently.</del>
Updated: Now  using `loaderContext.loaderIndex` in `NormalModule.js` for all to determine which loader the error is belong to .

### PR DESCRIPTION
**Pull request by mc-zone**
These additions resolve #2231 .

### CODE REVIEW

**Comment by montogeek** [comment]
@mc-zone Yes, you should, It was passing before.

**Comment by dhruvdutt** [comment]
@mc-zone Not sure why exactly it's failing but take a look at these [rules](https://github.com/DavidAnson/markdownlint/blob/master/doc/Rules.md#md031---fenced-code-blocks-should-be-surrounded-by-blank-lines).

**Comment by mc-zone** [comment]
>>>
> webpack.js.org@0.1.2 fetch /home/travis/build/webpack/webpack.js.org
> sh src/scripts/fetch.sh
{ [Error: {"message":"API rate limit exceeded for 35.184.48.144. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)","documentation_url":"https://developer.github.com/v3/#rate-limiting"}]
  message: '{"message":"API rate limit exceeded for 35.184.48.144. (But here\'s the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)","documentation_url":"https://developer.github.com/v3/#rate-limiting"}',
  code: 403,
  status: 'Forbidden',

Seems we have to fetch some content every time. Can these be cached or some other workaround? @montogeek

**Comment by montogeek** [comment]
@mc-zone I have tried, twice (#2270 , #1963), will try again

**Comment by mc-zone** [comment]
Anyone could review this? is this enough to be merged? 🚶

File: src/content/api/loaders.md [hunk 1]
```diff
@@ -471,3 +509,96 @@ Hacky access to the Compiler object of webpack.
...
+
+### Using `this.emitError`
+
+It just report errors but not interrupt module compiling.
+
+see [this.emitError](/api/loaders/#this-emiterror).
+
+### Using `throw` (or other uncaught exception)
+
+Throw out an uncaught error while loader are running will cause current module compiling failed.
```
Comment (montogeek) [inline src/content/api/loaders.md]: `...while a loader is running` is correct.
Correct word is `fail`.

File: src/content/api/loaders.md [hunk 2]
```diff
@@ -471,3 +509,96 @@ Hacky access to the Compiler object of webpack.
...
+
+```js
+
+module.exports = function(source){
+  throw new Error('Here is an fatal Error!');
+};
+
+```
+
+the module that ouccur this error will be packed into bundle (as an error module) like the following:
```
Comment (montogeek) [inline src/content/api/loaders.md]: Not sure what you mean on this line

File: src/content/api/loaders.md [hunk 3]
```diff
@@ -471,3 +509,96 @@ Hacky access to the Compiler object of webpack.
...
+
+ERROR in ./src/lib.js (./src/loader.js!./src/lib.js)
+Module build failed (from ./src/loader.js):
+Error: Here is an fatal Error!
+    at Object.module.exports (/workspace/src/loader.js:2:9)
+ @ ./src/index.js 1:0-25
+
+```
+
+And you can see these informations below, not only error message, but also details about which loader and module are involved:
```
Comment (montogeek) [inline src/content/api/loaders.md]: `As you can see below` would sound better

File: src/content/api/loaders.md [hunk 4]
```diff
@@ -471,3 +509,96 @@ Hacky access to the Compiler object of webpack.
 ### `this._module`
 
 Hacky access to the Module object being loaded.
+
+
+## Error Reporting
+
+There are some ways to throw out errors from inside a loader:
```
Comment (EugeneHlushko) [inline src/content/api/loaders.md]: this sounds abstract, please be more specific. e.g. `Here are the ways to throw an error from inside a loader:...`

File: src/content/api/loaders.md [hunk 5]
```diff
@@ -471,3 +509,96 @@ Hacky access to the Compiler object of webpack.
...
+
+### Using `this.emitError`
+
+It just report errors but not interrupt module compiling.
```
Comment (EugeneHlushko) [inline src/content/api/loaders.md]: `[this.emitError](/api/loaders/#this-emiterror) will report the errors without interrupting module's compilation.` or any other way you can rephrase this to sound closer to software documentation page

File: src/content/api/loaders.md [hunk 6]
```diff
@@ -471,3 +509,96 @@ Hacky access to the Compiler object of webpack.
...
+
+see [this.emitError](/api/loaders/#this-emiterror).
```
Comment (EugeneHlushko) [inline src/content/api/loaders.md]: see above

File: src/content/api/loaders.md (hunk 1 above)
Comment (EugeneHlushko) [reply src/content/api/loaders.md]: or: Throwing an error while loader is running will cause current module compilation failure.

File: src/content/api/loaders.md (hunk 2 above)
Comment (EugeneHlushko) [reply src/content/api/loaders.md]: The module on which loader had thrown this error will get bundled like this:

File: src/content/api/loaders.md (hunk 3 above)
Comment (EugeneHlushko) [reply src/content/api/loaders.md]: For sure, shorter and clearer

File: src/content/api/loaders.md [hunk 7]
```diff
@@ -471,3 +509,96 @@ Hacky access to the Compiler object of webpack.
...
+```
+
+And you can see these informations below, not only error message, but also details about which loader and module are involved:
+
+- the module path: `ERROR in ./src/lib.js`
+- the request string: `(./src/loader.js!./src/lib.js)`
+- the loader path: `(from ./src/loader.js)`
+- the caller path: `@ ./src/index.js 1:0-25`
+
+W> The loader path about error in the above are added after webpack 4.12
```
Comment (EugeneHlushko) [inline src/content/api/loaders.md]: The loader path in the error is displayed since webpack 4.12

**Comment by montogeek** [comment]
@mc-zone Friendly ping.

**Comment by mc-zone** [comment]
Updated.  @montogeek @EugeneHlushko Sorry about this...

File: src/content/api/loaders.md [hunk 8]
```diff
@@ -338,22 +338,46 @@ Should a source map be generated. Since generating source maps can be an expensi
...
+
+ERROR in ./src/lib.js (./src/loader.js!./src/lib.js)
+Module Error (from ./src/loader.js):
+Here is an Error!
+ @ ./src/index.js 1:0-25
+
 ```
 
-Emit an error.
+T> Unlike throwing an Error directly, it will NOT interrupt compiling process of current module.
```
Comment (EugeneHlushko) [inline src/content/api/loaders.md]: ...compilation process...

File: src/content/api/loaders.md [hunk 9]
```diff
@@ -362,7 +386,9 @@ Resolves the given request to a module, applies all configured loaders and calls
 ### `this.resolve`
 
 ``` typescript
+
```
Comment (EugeneHlushko) [inline src/content/api/loaders.md]: is this spacing required by the linter?

File: src/content/api/loaders.md [hunk 10]
```diff
@@ -471,3 +509,95 @@ Hacky access to the Compiler object of webpack.
...
+module.exports = function(source){
+  throw new Error('Here is an fatal Error!');
+};
+
+```
+
+The module on which loader had thrown this error will get bundled like this:
+
+```js-with-links
```
Comment (EugeneHlushko) [inline src/content/api/loaders.md]: do we still support js-with-links?

Comment (EugeneHlushko) [reply src/content/api/loaders.md]: cc @montogeek

Comment (montogeek) [reply src/content/api/loaders.md]: Yes

Comment (montogeek) [reply src/content/api/loaders.md]: This is master, it should work

Comment (montogeek) [reply src/content/api/loaders.md]: But you are not using them, `js` should be enough. Please change it.

Comment (EugeneHlushko) [reply src/content/api/loaders.md]: ok so there is only 1 outstanding issue i see with pr, compiling -> compilation

Comment (mc-zone) [reply src/content/api/loaders.md]: > But you are not using them, js should be enough. Please change it.

Here I want to display the original bundle code. but If I use `js`, the linter will scan the code and request for the fix. Could I ignore the linter here?

Comment (montogeek) [reply src/content/api/loaders.md]: What it the linter error? You could disable it.

File: src/content/api/loaders.md (hunk 9 above)
Comment (mc-zone) [reply src/content/api/loaders.md]: Yes it will report the error: `./src/content/api/loaders.md: 388: MD031 Fenced code blocks should be surrounded by blank lines`

**Comment by mc-zone** [comment]
@montogeek @EugeneHlushko  Hi, is there anything else that I should do for this PR? Please tell me if anything wrong. Thanks!

File: src/content/api/loaders.md [hunk 11]
```diff
@@ -471,3 +509,97 @@ Hacky access to the Compiler object of webpack.
...
+
+module.exports = function(source){
+  throw new Error('Here is a Fatal Error!');
+};
+
+```
+
+The module on which loader had thrown this error will get bundled like this:
+
+<!-- eslint-skip -->
```
[the remark this answers is not in the document]
Comment (mc-zone) [reply src/content/api/loaders.md]: Yes, it actually only works for the next code block, I confirmed.

**Comment by EugeneHlushko** [comment]
@montogeek please have another look

File: src/content/api/loaders.md [hunk 12]
```diff
@@ -25,7 +25,7 @@ Either `return` or `this.callback` can be used to return the transformed `conten
 
 __sync-loader.js__
 
-``` js
+``` javascript
```
Comment (jeremenichelli) [inline src/content/api/loaders.md]: Why just `js` doesn't work? 🤔

File: src/content/api/loaders.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (jeremenichelli) [reply src/content/api/loaders.md]: What is the reason we are disabling `eslint` here? And, isn't it possible to jsut disable the config that needs to be quiet down for the code block?

File: src/content/api/loaders.md (hunk 12 above)
Comment (EugeneHlushko) [reply src/content/api/loaders.md]: we are struggling from mixed declarations, since we use full language name for e.g. `typescript` or `bash` it is preferable to write full name here for consistency so that our *.md files are consistent and maintainable by any user coming from the internet :)

**Comment by montogeek** [comment]
Thanks!
